packages/teleport/src/teleportContextProvider.tsx

```diff
@@ -14,21 +14,67 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-import React from 'react';
+import React, { useEffect } from 'react';
+import styled from 'styled-components';
+import { Indicator } from 'design';
+import { Failed } from 'design/CardError';
+import useAttempt from 'shared/hooks/useAttemptNext';
 import TeleportContext from './teleportContext';
 
 // ReactContext is an instance of React context that is used to
 // access TeleportContext instance from within the virtual DOM
 const ReactContext = React.createContext<TeleportContext>(null);
 
-const TeleportContextProvider: React.FC<Props> = props => {
-  return <ReactContext.Provider value={props.ctx} children={props.children} />;
+export const TeleportContextProvider: React.FC<State> = props => {
```

Comment:
The intent of ContextProvider is to initialize react context provider with the value but with this change it's also responsible for rendering components (failed/loading/etc) which itself depend on the app layout. Lets discuss a better way (cleaner) to handle an initialization logic.

Reply:
I suppose implicit to this design is the idea that anywhere in the component hierarchy that's a child of this `TeleportContextProvider` we want:
- the context to be initialized
- if the context is loading, ignore all children and show a spinner (first screenshot below)
-  if the context initialization fails, show a `<Failed />` (second screenshot below)
<img width="1424" alt="Screen Shot 2022-02-11 at 09 17 15" src="https://user-images.githubusercontent.com/13578537/153607932-b9c10214-3a40-4c12-8179-e6bf41e574fb.png">
<img width="1439" alt="Screen Shot 2022-02-11 at 09 18 23" src="https://user-images.githubusercontent.com/13578537/153607953-719a3ce6-7988-489f-ad67-b7c1bf717cbc.png">

Presuming I'm correct about the behavior we want I can think of a few options for implementing that (or something reminiscent):
1. What I have in this PR
2. Rename the current [`useMain`](https://github.com/gravitational/webapps/blob/c0ea1b61d6cea778743b7a250f72a3a7aca86a52/packages/teleport/src/Main/useMain.ts#L22-L35) hook to something like `useInitializedContext`, call it in all of the [top level components below `TeleportContextProvider`](https://github.com/gravitational/webapps/blob/c0ea1b61d6cea778743b7a250f72a3a7aca86a52/packages/teleport/src/Main/useMain.ts#L22-L35), and then add the `ctx`-status-based [component rendering logic we currently have in `Main.tsx`](https://github.com/gravitational/webapps/blob/c0ea1b61d6cea778743b7a250f72a3a7aca86a52/packages/teleport/src/Main/Main.tsx#L36-L50) to each of those.
3. Keep the current `TeleportContextProvider` and add a `TeleportContextEnsurer` component which deals with the UI logic

I built this to avoid option 2, since it requires a lot of repetitive rendering logic and requires us to remember to implement it whenever we build a new top level component underneath `<TeleportContextProvider>`. Options 1 and 3 lead to a simpler system: if you're a child of `<TeleportContextProvider>`/<`TeleportContextEnsurer`>, you can always grab an initialized `TeleportContext`.

I can foresee a separation of concerns argument being made for going for something like option 3, however its not obvious to me that the concerns are separate. If we presume that the behavior I proposed is always desirable, then there's no need for the additional `Ensurer` distinction. I don't see why we should expect to need another `<TeleportContextProvider>` elsewhere in the app with a different rendering behavior. Granted I don't know anything about `teleterm`, and maybe its addition demands we design our `teleport` components with a greater eye towards shared usage. 

Let me know if I've erred in my reasoning or if there's a possibility I'm missing.